### DISCUSSION TITLE
Unable to create OTP-Seeds or recovery-tokens for Users by impersonation

### DISCUSSION THREAD

**Discussion by tim-at-gith** [keycloak/keycloak#34520] [Q&A]
I would like to give a user “manager” the possibility to create OTP seeds or recovery tokens for other users. So the user “manager” should be able to burn hardware OTP tokens for users or provide users with an "emergency access" if they have forgotten their hardware tokens.

The “manager” user has the impersonation right. 
If the “manager” user uses impersonation, he is also redirected to the user accout page and can customize the user data. 
However, if the user “manager” wants to create an OTP seed or Recoery token, a login is displayed for the user who is being edited.

Question: 
How do I configure keycloak so that the user “manager” can create OTP seeds and recovery tokens for other users without an additional login from the user?

By the way: 
Is there a way to ensure that it really is the user when the user logs in for the first time (without OTP/recovery tokens)? E.g. via a predefined PIN or Question/Answer that the user receives in a secure way?

**Comment by darius-m** [comment]
Would simply resetting the user's password not suffice? If you replace the user's tokens, they would be unable to access their account regardless, so resetting their password and adding an update password required action should be good enough in most cases. Also, if the tokens are set up as "required", the user should be asked to set up their tokens on their next login, so simply removing the tokens and asking the user to log in afterwards would allow the user to set up their tokens themselves.

For the last question, as far as I'm aware, no, there is no such option out of the box. You would need to use a custom plugin that generates the token, sends it over email and then removes it after being used. There are plugins for magic link logins and mail OTPs, as far as I know, so it should be doable. See the Extensions section from [the community tab on the keycloak site](https://www.keycloak.org/community) for some examples. However, the user's password should be proof that the user is who they claim to be, and if it's set as a temporary password that has to be reset on the first login, unless you assume it can be leaked before the first login, it should work as expected.

**Reply by tim-at-gith** [reply]
Thank you for the ideas

> Would simply resetting the user's password not suffice?

Not possible and also not wanted, because keycloak receives accounts from another source via user federation.

> Also, if the tokens are set up as "required", the user should be asked to set up their tokens on their next login, so simply removing the tokens and asking the user to log in afterwards would allow the user to set up their tokens themselves.

Deleting existing tokens is also not desired because they burned to hardware tokens. 
The second “problem” with this solution idea is that the creation of new tokens by users is restricted to special sources, e.g. not possible via the Internet.

> You would need to use a custom plugin that generates the token, sends it over email and then removes it after being used. There are plugins for magic link logins and mail OTPs, as far as I know, so it should be doable

I will have a look at magic link, mail OTPs will not work because mail access is protected with multifactor authentication.
